File: iOS7bookExamples/bk2ch07p378zoomDetailed/ch20p680zoomCentered/MyTiledView.m


#import "MyTiledView.h"

@interface MyTiledView ()
@property (strong, nonatomic) UIImage* currentImage;
@property (strong, nonatomic) NSValue* currentSize;
@end

@implementation MyTiledView

-(id)initWithCoder:(NSCoder *)aDecoder {
    self = [super initWithCoder:aDecoder];
    CATiledLayer* lay = (CATiledLayer*)self.layer;
    CGFloat scale = lay.contentsScale;
    lay.tileSize = CGSizeMake(208*scale,238*scale);
    lay.levelsOfDetail = 3;
    lay.levelsOfDetailBias = 2;
    self->_currentSize = [NSValue valueWithCGSize:CGSizeZero];
    return self;
}

+ (Class) layerClass {
    return [CATiledLayer class];
}

- (void)drawRect:(CGRect)rect
{
//    NSLog(@"rect %@", NSStringFromCGRect(rect));
//    NSLog(@"bounds %@", NSStringFromCGRect(self.bounds));
//    NSLog(@"contents scale %f", self.layer.contentsScale);
//    NSLog(@"%@", NSStringFromCGAffineTransform(CGContextGetCTM(UIGraphicsGetCurrentContext())));
//    NSLog(@"%@", NSStringFromCGRect(CGContextGetClipBoundingBox(UIGraphicsGetCurrentContext())));
    
          
    CGSize oldSize = self.currentSize.CGSizeValue;
    if (!CGSizeEqualToSize(oldSize, rect.size)) {
        // make a new size
        self.currentSize = [NSValue valueWithCGSize:rect.size];
        // make a new image
        CATiledLayer* lay = (CATiledLayer*) self.layer;
        
        CGAffineTransform tr = CGContextGetCTM(UIGraphicsGetCurrentContext());
        CGFloat sc = tr.a/lay.contentsScale;
        CGFloat scale = sc/4.0;
        
        NSString* path = [[NSBundle mainBundle] pathForResource:@"earthFromSaturn" ofType:@"png"];
        UIImage* im = [[UIImage alloc] initWithContentsOfFile:path];
        CGSize sz = CGSizeMake(im.size.width * scale, im.size.height * scale);
        UIGraphicsBeginImageContextWithOptions(sz, YES, 1);
        [im drawInRect:CGRectMake(0,0,sz.width,sz.height)];
        self.currentImage = UIGraphicsGetImageFromCurrentImageContext();
        UIGraphicsEndImageContext();
        NSLog(@"created image at size %@", NSStringFromCGSize(sz));
    }
    [self.currentImage drawInRect:self.bounds];
    
    // comment out the following! it's here just so we can see the tile boundaries
    
    UIBezierPath* bp = [UIBezierPath bezierPathWithRect: rect];
    [[UIColor whiteColor] setStroke];
    [bp stroke];

}

@end

File: bk2ch07p378zoomDetailed/ch20p680zoomCentered/MyTiledView.swift

import UIKit

func lend<T where T:NSObject> (closure:(T)->()) -> T {
    let orig = T()
    closure(orig)
    return orig
}


class MyTiledView : UIView {
    
    var currentImage : UIImage!
    var currentSize = CGSize.zero
    
    var drawQueue : DispatchQueue = DispatchQueue(label: "drawQueue", attributes: DispatchQueueAttributes.serial)
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder:aDecoder)
        let lay = self.layer as! CATiledLayer
        let scale = lay.contentsScale
        lay.tileSize = CGSize(208*scale,238*scale)
        lay.levelsOfDetail = 3
        lay.levelsOfDetailBias = 2
    }
    
    override class func layerClass() -> AnyClass {
        return CATiledLayer.self
    }
    
    override func draw(_ rect: CGRect) {
        //    NSLog(@"rect %@", NSStringFromCGRect(rect));
        //    NSLog(@"bounds %@", NSStringFromCGRect(self.bounds));
        //    NSLog(@"contents scale %f", self.layer.contentsScale);
        //    NSLog(@"%@", NSStringFromCGAffineTransform(CGContextGetCTM(UIGraphicsGetCurrentContext()!)));
        //    NSLog(@"%@", NSStringFromCGRect(CGContextGetClipBoundingBox(UIGraphicsGetCurrentContext()!)));
        
        
        
        drawQueue.sync { // work around nasty thread issue...
            // we are called twice simultaneously on two different background threads!

            let oldSize = self.currentSize
            // NSLog("oldSize %@", NSStringFromCGSize(oldSize))
            // NSLog("rect.size %@", NSStringFromCGSize(rect.size))
            if oldSize.equalTo(rect.size) {
                // NSLog("%@", "not equal, making new size")
                // make a new size
                self.currentSize = rect.size
                // make a new image
                let lay = self.layer as! CATiledLayer
                
                let tr = UIGraphicsGetCurrentContext()!.ctm
                let sc = tr.a/lay.contentsScale
                let scale = sc/4.0
                
                let path = Bundle.main.pathForResource("earthFromSaturn", ofType:"png")!
                let im = UIImage(contentsOfFile:path)!
                let sz = CGSize(im.size.width * scale, im.size.height * scale)
                
                
                let r = UIGraphicsImageRenderer(size: sz, format: lend {
                    (f : UIGraphicsImageRendererFormat) in
                    f.opaque = true
                    f.scale = 1
                })
                self.currentImage = r.image {
                    _ in
                    im.draw(in:CGRect(0,0,sz.width,sz.height))
                }

                
//                UIGraphicsBeginImageContextWithOptions(sz, true, 1)
//                im.draw(in:CGRect(0,0,sz.width,sz.height))
//                self.currentImage = UIGraphicsGetImageFromCurrentImageContext()!
//                UIGraphicsEndImageContext()
                
                NSLog("created image at size %@", NSStringFromCGSize(sz)) // only three times
            }
            self.currentImage.draw(in:self.bounds)
            
            // comment out the following! it's here just so we can see the tile boundaries
            
            let bp = UIBezierPath(rect: rect)
            UIColor.white().setStroke()
            bp.stroke()
            
        }
    }
}


